crates/sui-surfer/src/surfer_state.rs

```diff
@@ -0,0 +1,393 @@
+            .any(|arg| matches!(arg, CallArg::Object(ObjectArg::SharedObject { .. })));
+        let tx_data = TransactionData::new_move_call(
+            self.address,
+            package,
+            Identifier::new(module.as_str()).unwrap(),
+            Identifier::new(function.as_str()).unwrap(),
+            vec![],
+            self.gas_object,
+            args,
+            TEST_ONLY_GAS_UNIT_FOR_PUBLISH * rgp,
+            rgp,
+        )
+        .unwrap();
+        let tx = self.cluster.sign_transaction(&self.address, &tx_data);
+        let response = loop {
+            match self.cluster.execute_transaction(tx.clone()).await {
+                Ok(effects) => break effects,
+                Err(e) => {
+                    error!("Error executing transaction: {:?}", e);
+                    tokio::time::sleep(Duration::from_secs(1)).await;
+                }
+            }
+        };
+        debug!(
+            "Successfully executed transaction {:?} with response {:?}",
+            tx, response
+        );
+        let effects = response.effects.unwrap();
+        info!(
+            "[{:?}] Calling Move function {:?}::{:?} returned {:?}",
+            self.address,
+            module,
+            function,
+            effects.status()
+        );
+        self.stats.record_transaction(
+            use_shared_object,
+            effects.status().is_ok(),
+            package,
+            module,
+            function,
+        );
+        self.process_tx_effects(&effects).await;
+    }
+
+    async fn process_tx_effects(&mut self, effects: &SuiTransactionBlockEffects) {
+        for (owned_ref, write_kind) in effects.all_changed_objects() {
+            let obj_ref = owned_ref.reference.to_object_ref();
+            let object = self
+                .cluster
+                .get_object_from_fullnode_store(&obj_ref.0)
+                .await
+                .unwrap();
+            if object.is_package() {
+                self.discover_entry_functions(object).await;
+                continue;
+            }
+            let struct_tag = object.struct_tag().unwrap();
+            match owned_ref.owner {
+                Owner::Immutable => {
+                    self.immutable_objects
+                        .write()
+                        .await
+                        .entry(struct_tag)
+                        .or_default()
+                        .push(obj_ref);
+                }
+                Owner::AddressOwner(address) => {
+                    if address == self.address {
+                        self.owned_objects
+                            .entry(struct_tag)
+                            .or_default()
+                            .insert(obj_ref);
+                    }
+                }
+                Owner::ObjectOwner(_) => (),
+                Owner::Shared {
+                    initial_shared_version,
+                } => {
+                    if write_kind != WriteKind::Mutate {
```

Comment:
Wondering if write kind is not [Mutate](url), can it be anything other than `Create`? i.e. is it allowed to unwrap an object and share it?

Reply:
Not sure.

Reply:
Looked at the code. It seems possible

Reply:
https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/packages/sui-framework/sources/transfer.move#L55-L66
doesn't seem so?

Reply:
Ya.. you are probably correct and I'm probably wrong. 
https://github.com/MystenLabs/sui/blob/fe6ae8484d2df34f7a3b94f3f17ac83deb52fa64/crates/sui-move-natives/src/transfer.rs#L140
I guess `unwrap` is an owner change. What was the owner before? The object that wrapped it?

Reply:
Ok it seems that when an object is wrapped, it is considered deleted from the storage so there's no owner anymore. And when we put an object into storage via `share` or `transfer` or `freeze`, anything that isn't a new id creation or no owner change is considered `OwnerChanged` (including putting an unwrapped obj into storage). So `OwnerChanged` isn't a super accurate name, but fine for now.

Reply:
Then this line can probably be changed to `if write_kind == WriteKind::Create`